### PR TITLE
Replace React.error and React.warn with getComponentStack

### PR DESCRIPTION
**Pull request by bvaughn**
Proposed alternative to #15170

Several people have asked, "_Could you provide a method to access the component stack directly?_". I think our reasoning against this was that we didn't want people to _depend_ on that stack, or to try to parse it and build anything on top of it.

I don't think we considered the case of instrumenting error logging though. Facebook (and probably others) overrides `console.error` to capture the call stack and save it somewhere. A downside of `React.error` is that it would cause errors from different call sites to be grouped together (unless the instrumenting code was smart enough to skip over the React frame).

Overriding only these two console methods also feels like a bit of an incomplete solution (#15726) to the logging story.

This PR replaces `React.warn` and `React.error` with a method that returns the current component stack¹ which people can log to whichever console group they want.

¹ At the moment, this method always returns an empty string in prod.

### CODE REVIEW

File: packages/react/src/getComponentStack.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import ReactSharedInternals from 'shared/ReactSharedInternals';
+
+let getComponentStack = function() {
+  return '';
```
Comment (bvaughn) [inline packages/react/src/getComponentStack.js]: Should this not return anything? Seemed weird to return void/undefined in prod and an empty string when called outside of the render phase in DEV, so I opted to return empty string in both cases for now.

**Comment by sebmarkbage** [comment]
It’s not just about the component stack but other things we can do with it like reorder and delay etc.

This doesn’t address the abuse problem. The reason this wasn’t earlier is because there were many really bad use cases people were intending to use this for.

**Comment by bvaughn** [comment]
> It’s not just about the component stack but other things we can do with it like reorder and delay etc.

That's a topic for a broader logging strategy. I see it as separate from the more focused need of accessing a console stack. Do you disagree?

> This doesn’t address the abuse problem. The reason this wasn’t earlier is because there were many really bad use cases people were intending to use this for.

TBH bad actors would probably misuse e.g. `React.warn` as well by intercepting calls to `console.warn`. I think we can only advocate for **not**  doing that. On the other hand, the current approach (from #15170) doesn't account for existing error logging/aggregating use cases, which is the main reason I opened this PR in the first place...

**Comment by bvaughn** [comment]
Sebastian and I had a productive conversation this afternoon regarding my concerns above.

The quick takeaway is this: I will be updating this PR to remove `React.warn` and `React.error` _without_ adding an additional `getComponentStack` method. We can revisit these methods later as part of a larger discussion around the console logging experience with React (see #15726) but we should not block 16.9 for this.

In the meanwhile, I will be adding functionality to React DevTools to intercept calls to `console.error`, `console.warn`, and `console.trace` and auto-append component stacks (in DEV mode). This seems like a nice 80/20 solution that doesn't limit our options for a future API.

Might also be worth looking into a log view within DevTools that groups and orders logs by commit, etc.

**Comment by bvaughn** [comment]
Closing in favor of #16126

**Pull request closed without merging**